### PR TITLE
Migrate history correct db

### PR DESCRIPTION
**Pull request by prestwick**
- Created a new function for moving documents between databases
- refactored capture and restore into their own function
- added new arguments and logic to the new action type to allow intra-database migration
- Updated readme

- next steps tests, ideally with Mongo test harness.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -68,6 +68,15 @@ py systemlinkmigrate.py --capture --tag --dir="C:\migrate_8-3-2020
 ````
 These arguments will capture tag and tag history data and store them in the directory `C:\migrate_8-3-2020`
 
+## Migrating tag history to nitaghistorian database
+Due to a bug introduced in SystemLink 2020R2 it is possible for tag history data to be stored in the incorrect database within the MongoDB instance. **This bug only affects environments where a remote MongoDB server is used.** To determine if you are in this state connect a MongoShell or visual database tool such as [Mongo Compase](https://www.mongodb.com/products/compass) to your Mongo instance and check if the collections `metadata` and `values` are in the default authentication database (typically the `admin` database).
```
Comment (chris468) [inline README.md:72]: Tag history is the only one that is impacted when upgrading from R1, but is at least 1 other R2 service (security) that's also storing data in the wrong database.  Should the docs / param name cover all of it?

Might also be worth noting that a configuration change is necessary as well

File: slmigrate/mongohandler.py [hunk 2]
```diff
@@ -2,28 +2,61 @@
...
+        destination_collection = destination_db.get_collection(collection)
+        for document in source_collection:
+            print("Migrating " + str(document['_id']))
+            destination_collection.insert_one(document)
+
+
+def migrate_mongo_cmd(service, action, config):
+    migrate_within_instance(service, action, config)
+    capture_migration(service, action, config)
+    restore_migration(service, action, config)
```
Comment (chris468) [inline slmigrate/mongohandler.py]: Consider pulling the if statement out to this level.  Just reading this method, I was wondering why it first migrated within the instance, then captured and restored
```suggestion
    if action == constants.thdbbug.arg:
        migrate_within_instance(service, action, config)
    if action == constants.capture_arg:
        capture_migration(service, action, config)
    if action == constants.restore_arg:
        restore_migration(service, action, config)
```

File: slmigrate/mongohandler.py [hunk 3]
```diff
@@ -2,28 +2,61 @@
...
+    client = MongoClient(host=[no_sql_config[constants.no_sql.name]['Mongo.Host']], port=no_sql_config[constants.no_sql.name]['Mongo.Port'], username=no_sql_config[constants.no_sql.name]['Mongo.User'], password=no_sql_config[constants.no_sql.name]['Mongo.Password'])
+    source_db = client.get_database(name=constants.source_db, codec_options=codec)
+    destination_db = client.get_database(name=service.destination_db, codec_options=codec)
+
+    for collection in service.collections_to_migrate:
+        source_collection = source_db.get_collection(collection).find()
+        destination_collection = destination_db.get_collection(collection)
+        for document in source_collection:
+            print("Migrating " + str(document['_id']))
+            destination_collection.insert_one(document)
```
Comment (chris468) [inline slmigrate/mongohandler.py]: Does there need to be any logic as far as merging?  If there's metadata for a tag in both databases, does it need to be merged?  If there's history in both databases, is it ok to have multiple documents or do those need to be merged as well?

File: README.md (hunk 1 above)
Comment (prestwick) [reply README.md:72]: I agree. The `--thdbbug` argument is a placeholder for now. I'd like to have something more descriptive that covers the full scope of the problem. I'll make way for that in my next PR.  I'll also update the docs to reflect this as well, and include the configuration change we want people to make. I'll need your feedback on that to ensure we are putting our customer on the most successful path.

File: slmigrate/mongohandler.py (hunk 2 above)
Comment (prestwick) [reply slmigrate/mongohandler.py]: I will make this change. Does it seem worthwhile to keep the if statements within the functions as well? I was thinking I was making the function more "defensive" but if its just confusing I can pull it out.

Comment (prestwick) [reply slmigrate/mongohandler.py]: FYI 6af70737ab1cfa69e3df6bab96b7fca140206d7d

File: slmigrate/mongohandler.py (hunk 3 above)
Comment (prestwick) [reply slmigrate/mongohandler.py]: Good point. I added a try/except around this to look for duplicate documents. ff639f9b78265547d74a1fb75d9b968787514ef5 

What this doesn't accommodate for is when the document is different (different UUIDs), but what that document represents is the same (for example there was already a tag with the same path in this DB). I am guessing we could find ourselves in cases where after upgrading we start dumping data into the new/wrong database. while that metadata already exists in the correct one. In that case the merge is going to have to be more nuanced. I'll try and see if I can get something in to address this.

Comment (prestwick) [reply slmigrate/mongohandler.py]: After looking into this closer I am going to do the work for this merging scenario in another branch.

**Pull request merged**